cmd/neofs-node/reputation.go

```diff
@@ -334,20 +336,66 @@ func initReputationService(c *cfg) {
 	)
 }
 
-type loggingReputationServer struct {
-	log *logger.Logger
+type reputationServer struct {
+	log    *logger.Logger
+	cfg    *cfg
+	router trustcontroller.WriterProvider
+}
+
+type epochContext struct {
+	context.Context
+	epoch uint64
+}
+
+func (ctx *epochContext) Epoch() uint64 {
+	return ctx.epoch
+}
+
+type reputationOnlyKeyRemoteServerInfo struct {
+	key []byte
+}
+
+func (i *reputationOnlyKeyRemoteServerInfo) PublicKey() []byte {
+	return i.key
 }
 
-func (s *loggingReputationServer) SendLocalTrust(_ context.Context, req *v2reputation.SendLocalTrustRequest) (*v2reputation.SendLocalTrustResponse, error) {
+func (*reputationOnlyKeyRemoteServerInfo) Address() string {
+	return ""
+}
+
+func (s *reputationServer) SendLocalTrust(ctx context.Context, req *v2reputation.SendLocalTrustRequest) (*v2reputation.SendLocalTrustResponse, error) {
+	var passedRoute []reputationroute.ServerInfo
+
+	for hdr := req.GetVerificationHeader(); hdr != nil; hdr = hdr.GetOrigin() {
+		passedRoute = append(passedRoute, &reputationOnlyKeyRemoteServerInfo{
+			key: hdr.GetBodySignature().GetKey(),
+		})
+	}
+
+	for left, right := 0, len(passedRoute)-1; left < right; left, right = left+1, right-1 {
+		passedRoute[left], passedRoute[right] = passedRoute[right], passedRoute[left]
+	}
+
+	passedRoute = append(passedRoute, s)
+
 	body := req.GetBody()
 
-	log := s.log.With(zap.Uint64("epoch", body.GetEpoch()))
+	eCtx := &epochContext{
+		Context: ctx,
+		epoch:   body.GetEpoch(),
+	}
 
-	for _, t := range body.GetTrusts() {
-		log.Info("local trust received",
-			zap.String("peer", hex.EncodeToString(t.GetPeer().GetValue())),
-			zap.Float64("value", t.GetValue()),
-		)
+	w, err := s.router.InitWriter(reputationroute.NewRouteContext(eCtx, passedRoute))
+	if err != nil {
+		return nil, errors.Wrap(err, "could not initialize local trust writer")
+	}
+
+	trusts := apiToLocalTrust(body.GetTrusts())
```

Comment:
We iterate over `trusts`, so to avoid additional allocation we can iterate over `body.GetTrusts()` and convert single value per iteration.

Reply:
changed

Reply:
Nice, but lets return value instead of pointer since we dereference it anyway.

Reply:
changed